File: 0567-permutation-in-string/0567-permutation-in-string.cpp
class Solution {
public:
    bool isFrequencySame(int f1[26], int f2[26]) {
        for (int i = 0; i < 26; i++) {
            if (f1[i] != f2[i]) return false;
        }
        return true;
    }

    bool checkInclusion(string s1, string s2) {
        if (s1.length() > s2.length()) {
            return false;
        }

        int frequencyOfS1[26] = {0};
        int windowFrequency[26] = {0};

        // Calculate the frequency of s1
        for (int i = 0; i < s1.length(); i++) {
            int ascii = s1[i] - 'a';
            frequencyOfS1[ascii]++;
        }

        int windowSize = s1.length();
        for (int i = 0; i < s2.length(); i++) {
            // Add the current character to the window frequency
            windowFrequency[s2[i] - 'a']++;

            // Remove the character that is sliding out of the window
            if (i >= windowSize) {
                windowFrequency[s2[i - windowSize] - 'a']--;
            }

            // Compare frequencies only when the window size matches s1
            if (i >= windowSize - 1 && isFrequencySame(frequencyOfS1, windowFrequency)) {
                return true;
            }
        }

        return false;
    }
};

File: 0567-permutation-in-string/0567-permutation-in-string.java
class Solution {

    static private boolean check(int[] freqnS1, int[] freqnS2){
        for(int i = 0; i < 26; i++){
            if(freqnS1[i] != freqnS2[i]) return false;
        }
        return true;
    }
    public boolean checkInclusion(String s1, String s2) {
        int window = s1.length();
        int[] freqnS1 = new int[26];
        int[] freqnS2 = new int[26];

        for(char c: s1.toCharArray()){
            freqnS1[c - 'a']++;
        }

        for(int i = 0; i < s2.length(); i++){
            freqnS2[s2.charAt(i) - 'a']++;

            if(i >= window){
                freqnS2[s2.charAt(i - window) - 'a']--;
            }

            if(check(freqnS1, freqnS2)){
                return true;
            }
        }
        return false;
    }
}